File: merge sort/merge_sort.py
def Merge(arr, start, middle, end):
	"""Merging two arrays by ascending order."""
	front = middle - start + 1
	rear = end - middle
	left = [arr[i+start] for i in range(front)]
	right = [arr[i+middle+1] for i in range(rear)]
	l = 0
	r = 0

	for i in range(start, end+1):
		if l < front and r < rear:
			if left[l] < right[r]:
				arr[i] = left[l]
				l += 1
			else:
				arr[i] = right[r]
				r += 1
		else:
			if l < front:
				arr[i] = left[l]
				l += 1
			else:
				arr[i] = right[r]
				r += 1

				
def Show(arr):
	"""Display all elements in an array."""
	for i in range(len(arr)):
		print(str(arr[i]) if i == len(arr)-1 else str(arr[i]) + ' ', end='')


def Sort(arr, start, end):
	"""Sorting array to two arrays."""
	if start < end:
		middle = int((start + end) / 2)
		Sort(arr, start, middle)
		Sort(arr, middle+1, end)
		Merge(arr, start, middle, end)


n = int(input())
numbers = list(map(int, input().split()))
Sort(numbers, 0, n-1)
Show(numbers)

File: merge sort/merge_sort.c
#include <stdio.h>

void Merge(int *arr, int *start, int *middle, int *end);
void Show(int *arr, int *size);
void Sort(int *arr, int start, int *end);

int main()
{
	int n = 0;
	scanf("%d", &n);
	int numbers[n];

	for (int i = 0; i < n; i++) {
		scanf("%d", &numbers[i]);
	}

	int end = n-1;
	Sort(numbers, 0, &end);
	Show(numbers, &n);
	return 0;
}

void Merge(int *arr, int *start, int *middle, int *end) {
	int front = *middle - *start + 1;
	int rear = *end - *middle;
	int left[front], right[rear];

	for (int i = *start; i <= *end; i++) {
        if (i <= *middle) {
            left[i - *start] = arr[i];
        } else {
          right[i - *middle-1] = arr[i];
        }
	}

	for (int i = *start, l = 0, r = 0; i <= *end; i++) {
		if ((l < front) && (r < rear)) {
			arr[i] = left[l] < right[r] ? left[l++] : right[r++];
		} else {
			arr[i] = l < front ? left[l++] : right[r++];
		}
	}
}

void Show(int *arr, int *size) {
	for (int i = 0; i < *size; i++) {
		i == *size - 1 ? printf("%d", arr[i]) : printf("%d ", arr[i]);
	}
	printf("\n");
}

void Sort(int *arr, int start, int *end) {
	if (start < *end) {
		int middle = (start + *end) / 2;
		Sort(arr, start, &middle);
		Sort(arr, middle+1, end);
		Merge(arr, &start, &middle, end);
	}
}